### PR TITLE
Begin templating Eloquent

### PR DESCRIPTION
**Pull request by mr-feek**
👋 Continuing to iterate on getting one of my local laravel projects to pass level 6. Lot's more to go

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
         "illuminate/database": "5.8.* || ^6.0 || ^7.0",
         "illuminate/http": "5.8.* || ^6.0 || ^7.0",
         "illuminate/support": "5.8.* || ^6.0 || ^7.0",
-        "vimeo/psalm": "^3.8.2",
+        "vimeo/psalm": "^3.8.2|dev-master",
```
Comment (mr-feek) [inline composer.json]: Is this cool?

File: tests/acceptance/PhpstormGenerics.feature [hunk 2]
```diff
@@ -0,0 +1,37 @@
+Feature: Phpstorm Generic Support
```
Comment (mr-feek) [inline tests/acceptance/PhpstormGenerics.feature:1]: Unrelated to this PR at hand, but I was curious if it was working or not. I included it here 🤷

File: composer.json (hunk 1 above)
Comment (muglug) [reply composer.json]: Sure

File: src/ReturnTypeProvider/ModelReturnTypeProvider.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        $was_inside_call = $context->inside_call;
+
+        $context->inside_call = true;
+
+        // call here
+        // todo: right now  this is just for static
+
+        $fakeMethodCall = new StaticCall(new Name(Builder::class), $method_name_lowercase, $call_args);
```
Comment (mr-feek) [inline src/ReturnTypeProvider/ModelReturnTypeProvider.php]: @muglug I'm struggling with this. How do we fake a method call to a generic object? How do I pass the template parameters? Would love any tips you may have!

Essentially I want to say that  `__call` is a proxy to `Builder<User>`

Comment (mr-feek) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php]: Meanwhile working on a stubbing approach...

Comment (muglug) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php]: Mind opening a ticket in vimeo/psalm for that?

File: src/ReturnTypeProvider/ModelReturnTypeProvider.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+use Psalm\Plugin\Hook\MethodExistenceProviderInterface;
+use Psalm\Plugin\Hook\MethodReturnTypeProviderInterface;
+use Psalm\StatementsSource;
+use Psalm\Type;
+use Psalm\Type\Atomic\TGenericObject;
+use Psalm\Type\Atomic\TNamedObject;
+use Psalm\Type\Atomic\TTemplateParam;
+use Psalm\Type\Union;
+
+final class ModelReturnTypeProvider implements MethodReturnTypeProviderInterface
```
Comment (SCIF) [inline src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: @muglug , @mr-feek , do you think implementation of such provider is better than fix #3219 ?

Comment (mr-feek) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: I do not think this is a good implementation approach -- I'm trying to solve by templating. Just using  this draft as a brainstorming area..

Comment (mr-feek) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: This was how we solved it over at larastan, but proxying for return types is much easier over there. If we can solve via templating, I think we should strive for that appraoch!

Comment (SCIF) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: Correct me if I'm wrong but it's easy task: `@return static&Builder<static>` for several model methods. WDYT?

Comment (mr-feek) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: Yeah, I've been debating on just hardcoding `Builder` vs figuring out the actual type of builder etc. Feel free to make a PR if you have ideas!

Comment (mr-feek) [reply src/ReturnTypeProvider/ModelReturnTypeProvider.php:17]: I think the easiest is to just copy/paste the builder methods and put them in the model stub. I've been messing with rewriting the docblocks that ide helper generates to make them correct.

File: tests/acceptance/EloquentBuilderTypes.feature [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        /**
+        * @return \Illuminate\Database\Eloquent\Builder<User>
+        */
+        public function getNewQuery(): \Illuminate\Database\Eloquent\Builder
+        {
+          return User::query();
+        }
+
+        /**
+        * todo: i think it's a psalm bug that we have to &static, but I don't know enough about it.
```
Comment (mr-feek) [inline tests/acceptance/EloquentBuilderTypes.feature]: https://github.com/vimeo/psalm/issues/3296

**Comment by mr-feek** [comment]
This approach is pretty much blocked by  https://github.com/vimeo/psalm/issues/3294 and https://github.com/vimeo/psalm/issues/3295

**Comment by muglug** [comment]
Sorry - I’ll hopefully get to those tomorrow

**Comment by mr-feek** [comment]
Tests are green, gonna  test on some local projects  of mine

**Pull request merged**